windows.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2017 The CRT Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:generate go run generator.go
+
+// +build windows
+
+package crt
+
+import (
+	"fmt"
+	"os"
+	"sync/atomic"
+	"syscall"
+	"unsafe"
+
+	"github.com/cznic/memory"
+)
+
+// TODO: implement a generic wide string variant of this
+// GoUTF16String converts a wide string to a GOString using
+// windows-specific implementations in go's syscall package
+func GoUTF16String(s uintptr) string {
+	ptr := (*[1 << 20]uint16)(unsafe.Pointer(s))
+	return syscall.UTF16ToString(ptr[:])
+}
+
+// DWORD WINAPI GetLastError(void);
+func XGetLastError(tls TLS) uint32 {
+	return uint32(tls.err())
+}
+
+// DWORD WINAPI GetCurrentThreadId(void);
+func XGetCurrentThreadId(tls TLS) uint32 {
+	return uint32(tls.getThreadID())
+}
+
+func X_msize(tls TLS, ptr uintptr) uint32 {
+	return uint32(memory.UintptrUsableSize(ptr))
+}
+
+func X_endthreadex(tls TLS, a uint32) {
+	panic("TODO not implemented and not used")
+}
+
+func X_beginthreadex(tls TLS, a uintptr, b uint32, c uintptr, d uintptr, e uint32, f uintptr) uintptr {
+	panic("TODO not implemented and not used")
+}
+
+// LONG __cdecl InterlockedCompareExchange(_Inout_ LONG volatile *Destination,_In_ LONG Exchange,_In_ LONG Comparand);
+// TODO: figure out if we can bypass a minor race (see below for an explanation)
+func X_InterlockedCompareExchange(tls TLS, dest_param uintptr, exchange, comparand int32) int32 {
+	// TODO: memory barrier: https://msdn.microsoft.com/de-de/library/windows/desktop/ms683560(v=vs.85).aspx
+
+	if strace {
+		fmt.Fprintf(os.Stderr, "InterlockedCompareExchange(%#x, %#x, %#x)\n", comparand, exchange, dest_param)
+	}
+
+	dest := (*int32)(unsafe.Pointer(dest_param))
+
+	initial := comparand
+	if !atomic.CompareAndSwapInt32(dest, comparand, exchange) {
+		initial := *dest
+		if initial == comparand {
+			// we cannot prevent all cases of races using this implementation, since we have to
```

Comment:
Please add a `//TODO` line, assuming there's something we can do about it in the future.

Reply:
The TODO is at the top of the function

Reply:
Great, then it's okay.